### PR TITLE
first shot at adding ~coveralls~ codecov integration

### PR DESCRIPTION
**Pull request by amueller**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -175,6 +175,7 @@ RUN python3.7 -m pip install pip && \
 COPY ./source/Mlos.Python/requirements.txt /tmp/
 RUN python3.7 -m pip install -r /tmp/requirements.txt
 
+
```
Comment (bpkroth) [inline Dockerfile:178]:
```suggestion
```

Comment (bpkroth) [reply Dockerfile:178]: Or do we need `pip install pytest-cov`?

Comment (amueller) [reply Dockerfile:178]: yes

**Comment by amueller** [comment]
right now the PR coverage diff isn't working because there's no coverage for the base branch. So I think we should merge so that we can get coverage for that?

I have to see how to enable it as a CI check, but we already have file-level coverage for this PR: https://codecov.io/gh/microsoft/MLOS/pull/187/tree?path=source%2FMlos.Python%2Fmlos

**Comment by amueller** [comment]
Not sure if you can see it, but in the settings here:
https://app.codecov.io/gh/microsoft/MLOS/settings

it tells me to enable something in the MS org which I don't have access to :-/

**Comment by bpkroth** [comment]
> 
> 
> Not sure if you can see it, but in the settings here:
> https://app.codecov.io/gh/microsoft/MLOS/settings
> 
> it tells me to enable something in the MS org which I don't have access to :-/

No, I get a "sorry you don't have access to this", but I didn't try very hard ...

We should check with @ksaur about what they did for the hummingbird repo.

**Comment by bpkroth** [comment]
Related: #33 (though perhaps not all of it)

**Comment by ksaur** [comment]
@bpkroth  Hmm....I don't remember having to enable something special.  Ping me tomorrow and we can try to troubleshoot.  (I just checked that I can verify that link above if I `s/MLOS/hummingbird/g`, but I'm not sure what I did to enable it.)

**Comment by amueller** [comment]
btw looks like we don't need to have a yaml file:
https://docs.codecov.io/docs/codecov-yaml

When I go to the config page, at the top it says "Github Integration is installed. However, this repository is not enabled.".

**Comment by amueller** [comment]
I still think we should merge this as a first pass btw.

**Comment by bpkroth** [comment]
Do you want to add a badge to the front page, or leave that for later?
https://app.codecov.io/gh/microsoft/MLOS/settings/badge

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -536,6 +536,11 @@ jobs:
     - name: Run Python unit tests (Ubuntu ${{ matrix.UbuntuVersion }})
       timeout-minutes: 30
       run: docker exec mlos-build-python-${{ matrix.UbuntuVersion }} make python-test
+    - name: Upload coverage to codecov
+      uses: codecov/codecov-action@v1
+      with:
+        file: ./coverage.xml
+        flags: unittests
```
Comment (bpkroth) [inline .github/workflows/main.yml:543]:
```suggestion
        flags: unittests
        fail_ci_if_error: true
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -536,6 +536,11 @@ jobs:
     - name: Run Python unit tests (Ubuntu ${{ matrix.UbuntuVersion }})
       timeout-minutes: 30
       run: docker exec mlos-build-python-${{ matrix.UbuntuVersion }} make python-test
+    - name: Upload coverage to codecov
```
Comment (bpkroth) [inline .github/workflows/main.yml:539]:
```suggestion
    - name: Upload coverage to codecov
      if: ${{ github.repository == 'microsoft/mlos' }}
```

Comment (bpkroth) [reply .github/workflows/main.yml:539]: To prevent uploads when the actions are run from a forked repo during testing.

**Comment by amueller** [comment]
added the badge: https://github.com/microsoft/MLOS/blob/97249bc5a7ef467f8eeaeedf8cc262e9bb1e3e56/README.md

File: .github/workflows/main.yml (hunk 3 above)
Comment (bpkroth) [reply .github/workflows/main.yml:539]: Although, would those just go to their own codecov collection?  If so, maybe we don't care and that's useful for testing as well, so nevermind.

File: .github/workflows/main.yml (hunk 2 above)
Comment (bpkroth) [reply .github/workflows/main.yml:544]: Undecided about this.  Seems more useful if we're providing a badge, but then we're dependent upon an external service.

**Review by bpkroth — APPROVED** [review]
Thanks!

**Pull request merged**